### DISCUSSION TITLE
Should we support temporal elements?

### DISCUSSION THREAD

**Discussion by stephenwf** [digirati-co-uk/iiif-canvas-panel#21] [Q&A] [answered]
In IIIF there are 2 types of temporal elements. Single resources and compositions of resources. A single resource may be an single audio or video painted onto a canvas with a duration matching. In viewers this is usually treated as a special case and put into a dedicated audio or video player. The second type of temporal element is anything beyond this simple case. For example, multiple audio tracks painted onto a canvases temporal dimension, played back in a different order, or jumping between the two tracks or IIIF images painted onto the temporal dimension to show/hide at various time stamps.

The data layers should already support the former types, so the question is wether we should build fundamental support for temporal elements into the rendering (Atlas).

## Web component example
To demonstrate what we could enable with a temporal-aware viewer, here's some aspirational markup for describe a simple canvas.
```html
<atlas>
  <!-- world object can be used to represent a IIIF canvas -->
  <world-object id="obj1" duration="60" height="500" width="500">
    
    <img src="http://example.org/image-1.jpg" t="0, 30" x="0" y="0" width="500" height="500" />
    <!--                                     ^--------^                           -->
    <!--          This describes where this image should be displayed in time.    -->

    <img src="http://example.org/image-2.jpg" t="30, 60" x="0" y="0" width="500" height="500" />
    
  </world-object>
</atlas>
```
So we've declared a canvas with 2 images and a 60 second duration. The first shows for 30 seconds, and then the second image for 30 seconds. Internally the viewer will simply look at the current time and based on that decide if it needs to render a particular item. 

You could imagine getting a hold of the "clock" for this canvas:

```js
const obj1 = document.getElementById('obj1');

obj1.play();
obj1.pause();
obj1.seekTo(35);
obj1.stop();

obj1.duration; // 60.
obj1.currentTime; // 1.23556

```
and binding these to some external UI or a scrubber, as though it was a native video element.

**In this way a canvas inside of Atlas could be interacted in the same way as an HTML audio or video element.**

## How?

In order to get to a point where a viewer can be aware of an internal clock and render based on it, we need to have a way to first represent arbitrary things (like images appearing and disappearing) as if they were media elements that can be played/paused etc. We also need a way to compose these elements together, alongside _real_ media elements, such as audio or video elements.

![image](https://user-images.githubusercontent.com/8266711/128367762-33d37ab3-4c7c-4258-bcaf-0fa5d5041607.png)

* **Media** - We need a way to represent the media we want to use and we want them to all look and work the same. They would have start and end times, and be playable. 
*  **Track** - Once we have our media, we want a place to compose it together. Our track may be a IIIF canvas, where a piece of media is "painted" onto the track at a specific time. (e.g. play 0-30s of media-1 onto the track starting at 40s). When a track is played, it may jump from one media to another.
* **Sequence** - Lots of media interfaces have a concept of a playlist or many tracks that play in sequence. This drives next/previous functionality or track listing. This would be a way to model this, while still having a playable object (duration, current time, play/pause etc.)
* **Composition** - Lastly is a composition. Unlike a sequence a composition may grab parts of a track, or play them out of order. In IIIF this may appear as a set of ranges targeting timestamps on canvases. (Canvases = tracks). If you were to conceptually "play" a range, it could be represented as a composition. Like the other objects, it would be playable (duration, current time, play/pause)

### Challenges

There are some challenges with composing multiple elements together, including potentially browser-native elements such as audio. 

**Clock synchronisation**
The primary concern is keeping clocks in sync. When composing a track with multiple media elements, we really need to keep an internal clock ticking with what we consider the "current time". We then need to make sure that the media elements are keeping time with us. However, if you do this with zero drift the HTML audio would stutter constantly. Allowing for the clocks to go half a second or so out of phase and correcting if they do is important. For this reason, each of the elements described above need a way to be told a canonical time and internally synchronise to that time. 

**Buffering**
When relying on multiple pieces of media, especially when playing out of order, there is likely to be buffering of audio. This means at any time, any of the media elements you have composed together can stop the internal clock. This is closely related to clock synchronisation.

**Media controls**
If composing native elements, the operating system can pause or play those elements through the media keys. This means that any native element not only is controlled by our clock synchronisation, but can also affect the internal clock by stopping and starting it. The key is to avoid an infinite loop where media tells the clock, which then tells in turn tells the media and then tells the clock - to infinity. 

## Tools + libraries

To enable this functionality there are a few experimental tools and extensions that could be planned and worked on. The first is a tool to compose all of these layers of media together (real or "fake") with a standard set of APIs for playback. Any user interface written against this API should be able to render play/pause, duration, scrubber controls for anything from the simplest to most complex case.

The second is native support for temporal elements in Atlas. So this involved being able to pass down a duration to objects, and declaring objects as only being displayed between a start/end point. If we set up Atlas to construct media-kit objects, these deep zoom viewers can be played back, composed as if they were a normal audio/video element.

### Media kit

[repository](https://github.com/atlas-viewer/media-kit)

A completely non-IIIF specific library has been drawn up, based on the concepts above and trying to solve the synchronisation issues. A lot of the prior work for related AV work (From uses cases like the British Library) has been distilled here.

```js
// Source audio element.
const audio = new MediaKit.HTMLAudio('https://actions.google.com/sounds/v1/alarms/bugle_tune.ogg', {});

// Wrapper into media (optional)
const media = new MediaKit.Media({
    adapter: audio,
    duration: 6.233
});

// Creating a track - i.e. a canvas
const track = new MediaKit.Track();
track.addMedia(audio, {
    source: {
        start: 1,
        end: 4
    },
    target: {
        start: 2,
        end: 5,
    }
});

// Creating a player instance - this has the internal clock
// and play/pause/scrubbing/duration/current time for building a UI
// Some elements, like the audio here, don't need a player to work and have play/pause 
// on their objects. However, something like an Image showing/hiding does not have an 
// internal clock, so a player is required. Player will also keep multiple audio tracks synchronised.
const player = new MediaKit.Player(media);
```

### Native Atlas support
As mentioned in the first snippet with web components, if Atlas can understand durations and either contain an internal clock (likely provided by Media kit) or be provided one by a user, then that opens up Atlas to render almost everything it can already render but with time targets.

**Comment by tomcrane** [comment]
Worth re-reading Joscha's notes here: https://github.com/digirati-co-uk/iiif-av-bl and https://github.com/digirati-co-uk/iiif-av-bl/blob/master/EVALUATION.md

This was 4+ years ago and browsers evolve.
https://developer.mozilla.org/en-US/docs/Web/API/Web_Audio_API/Using_Web_Audio_API

One problem is equivalence of timing issues across audio and video sources - we want it to work the same for both. Does Media Kit solve that?

**Reply by stephenwf** [reply]
Yes that's one of the main problems. The way the AV component ended up working was having an internal clock and some allowed drift of audio/video sources. This will result in very slight pauses if any element drifts too far. This wouldn't be enough for music production but should be accurate enough for playing back. 

I think a lot of the timing APIs have been scaled back. High resolution clocks have been scaled back in browsers, with jitter artificially added. Ideally there would be a way to let the browser synchronise - but until then we need to do it manually. The web audio api link you posted _might_ be able to merge audio tracks, but because you can access the byte data of the audio, I expect it would require CORS on all audio links.

This is the synchronisation, distilled from the AV component.
https://github.com/atlas-viewer/media-kit/blob/main/src/dom/html-audio.ts#L101-L105

**Comment by tomcrane** [comment, marked as answer]
For the first pass at Canvas Panel, we shouldn't attempt to solve all the above - we should focus on 2D rendering use cases.

However, we can handle the 99% case - where the canvas has one AV content source, and the `duration` matches (within permitted slop) the duration of the content. In this special (but 99.99..%) case, we just turn the content into an HTML5 `<audio>` or `<video>` tag, set the source elements to the Choice resources (if present), and _possibly_ handle text content, if time allows.
